typings/index.d.ts

```diff
@@ -1,1567 +1,31 @@
-
-  id: string;
-
-  options: BaseSceneOptions<TContext>;
-
-  enterHandler: Middleware<TContext>;
-
-  leaveHandler: Middleware<TContext>;
-
-  ttl?: number;
-
-  enter: (...fns: Middleware<TContext>[]) => this;
-
-  leave: (...fns: Middleware<TContext>[]) => this;
-
-  enterMiddleware: () => Middleware<TContext>;
-
-  leaveMiddleware: () => Middleware<TContext>;
-}
-
-export type Scene<TContext extends SceneContextMessageUpdate> = BaseScene<TContext>;
-
-export type StageOptions = SceneContextOptions;
-
-export class Stage<TContext extends SceneContextMessageUpdate> extends Composer<TContext> {
-  constructor(scenes: Scene<TContext>[], options?: Partial<StageOptions>)
-
-  register: (...scenes: Scene<TContext>[]) => this;
-
-  middleware: () => Middleware<TContext>;
-
-  static enter: (sceneId: string, initialState?: object, silent?: boolean) => Middleware<SceneContextMessageUpdate>;
-
-  static reenter: () => Middleware<SceneContextMessageUpdate>;
-
-  static leave: () => Middleware<SceneContextMessageUpdate>;
-}
-
-export function session<TContext extends ContextMessageUpdate>(opts?: Partial<{
-  property: string;
-  store: Map<string, any>;
-  getSessionKey: (ctx: TContext) => string;
-  ttl: number;
-}>): Middleware<TContext>;
-
-export interface TelegrafConstructor extends ComposerConstructor {
-  /**
-   * Initialize new Telegraf app.
-   * @param token Bot token
-   * @param options options
-   * @example
-   * new Telegraf(token, options)
-   */
-  new <TContext extends ContextMessageUpdate>(token: string, options?: TelegrafOptions): Telegraf<TContext>;
-
-  log(logFn?: Function): Middleware<ContextMessageUpdate>;
-}
-
-export interface TOptions {
-
-  /**
-   * Telegram options
-   */
-  telegram?: {
-    /**
-     * https.Agent or http.Agent instance, allows custom proxy, certificate, keep alive, etc.
-     */
-    agent: https.Agent | http.Agent
-
-    /**
-     * Reply via webhook
-     */
-    webhookReply: boolean
+import { BaseScene, Stage } from './stage'
+import { Composer, Middleware } from './composer'
+import { ContextMessageUpdate } from './context'
+import { Extra } from './extra'
+import { Markup } from './markup'
+import { Router } from './router'
+import { session } from './session'
+import { Telegram } from './telegram'
+
+// works around https://github.com/microsoft/TypeScript/issues/37712
+import { Telegraf as _Telegraf } from './telegraf'
+declare class Telegraf<TContext extends ContextMessageUpdate> extends _Telegraf<
+  TContext
+> {}
+
+declare namespace Telegraf {
+  export {
+    BaseScene,
+    Composer,
+    ContextMessageUpdate as Context,
+    Extra,
+    Markup,
+    Middleware,
+    Router,
+    session,
+    Stage,
+    Telegram,
   }
-
-  /**
-   * Bot username
-   */
-  username?: string
-
-  /**
-   * Handle `channel_post` updates as messages
-   */
-  channelMode?: boolean
-
-  retryAfter?: number
-
-  handlerTimeout?: number
 }
 
-export default Telegraf
+export = Telegraf
```

Comment:
This would unnecessarily export a lot of internal types. I curated `index.d.ts` to only export types for stuff exported by `telegraf.js`, and `Middleware`.
Also, the workaround is not needed in your code since there's no declaration merging.

Reply:
I dont only create bots, I also create middlewares for Telegraf which sometimes need internal types. Also things like Markup contain a lot of types which have to be exported. Having to list them each manually here to is prone to errors.

Reply:
Also: whats the problem with exporting types and interfaces? Typescript is there to ensure type safety. Not being able to use types because they are inaccessible is TypeScript as useless as JavaScript. When they are exported and you dont need them you still have the option to not import them.

Reply:
It clutters suggestions, making the library look more intimidating that it actually is.

Reply:
> using `export = Telegraf` will require all repositories which use Telegraf to include `esModuleInterop` in the tsconfig.

I am `require`'ing the module with improved typings without it, no issues.
Also, see https://github.com/telegraf/telegraf/issues/763#issuecomment-538671190 and #943 -- situation is similar to `Context`.

Reply:
I think the `export default Telegraf` is correct as long as a library has a `module.export.default` (if i understand [this](https://github.com/DefinitelyTyped/DefinitelyTyped#a-package-uses-export--but-i-prefer-to-use-default-imports-can-i-change-export--to-export-default) correctly). The default was added in #445 to use it as a default import.

Based on that im not sure the default export is what is useful here and probably also confusing VSCode Intellisense.

Doing a step back from Telegraf and looking at the way it is done elsewhere. In JavaScript / CommonJS I would use a module like this: 
```js
const fs = require('fs')
fs.writeFileSync(…)
```

ES6 modules are more into this way of doing things:
```js
import * as fs from 'fs'
fs.writeFileSync(…)
// or
import {writeFileSync} from 'fs'
writeFileSync(…)
```

Back to Telegraf. Seeing this I would assume I should import it like this?
```js
import {Telegraf, Markup} from 'telegraf'
// or
import * as Telegraf from 'telegraf'
```

When there is a default export like we have here I should be able to do this:
```js
import Telegraf, {Markup} from 'telegraf'
```

Seeing that i would assume there is a default export besides the other exports like this:
```js
export default Telegraf;
export {Markup} from …;
```

Also I don't understand why the namespace is there. We are already in the TypeScript module 'Telegraf' so why is there an additional namespace for that? What is its benefit?
Taking a look at the TypeScript Module Documentation it also lists top level namespaces as a [red flag](https://www.typescriptlang.org/docs/handbook/modules.html#red-flags).

Reply:
Thank you for your persistence on this.
Since it's impossible to provide maximally convenient typings for both CommonJS and TypeScript consumers, I deprioritized CommonJS and significantly simplified my typings as a result. I also added `Telegraf` and `ContextMessageUpdate` aliases for backwards compatibility.

Reply:
Looking on other JavaScript projects like express with typings added via @types packages they mostly are used this way:

```ts
import express as * from 'express'
const app = new express()
```

DefinitelyTyped is relatively strict about not using a default export.

But as Telegraf already had a default export for years I think the state we are in with this PR is quite good now.